File: ios/BudgetBuddy/ViewModels/ReportsViewModel.swift
import Foundation

struct CategorySpend: Identifiable {
    var id   = UUID()
    let name: String
    let icon: String
    let amount: Double
    let colorHex: String
}

struct MonthTotal: Identifiable {
    var id = UUID()
    let label: String
    let total: Double
}

struct CategoryBudgetBar: Identifiable {
    var id = UUID()
    let categoryName: String
    let icon: String
    let colorHex: String
    let spent: Double
    let minAmount: Double
    let limitAmount: Double
}

struct ReportsUiState {
    var balance: Double              = 0
    var totalExpense: Double         = 0
    var totalIncome: Double          = 0
    var transactions: [BBTransaction] = []
    var categorySpends: [CategorySpend] = []
    var categoryBudgetBars: [CategoryBudgetBar] = []
    var monthlyTotals: [MonthTotal]  = []
    var isLoading: Bool              = false
}

@MainActor
final class ReportsViewModel: ObservableObject {
    @Published var uiState = ReportsUiState()
    @Published var selectedMonth: Int  = Calendar.current.component(.month, from: .now) - 1  // 0-based
    @Published var selectedYear: Int   = Calendar.current.component(.year,  from: .now)

    private let transactionRepo: TransactionRepository
    private let budgetRepo: BudgetRepository
    private let categoryRepo: CategoryRepository
    private var userId = ""

    init(transactionRepo: TransactionRepository, budgetRepo: BudgetRepository,
         categoryRepo: CategoryRepository) {
        self.transactionRepo = transactionRepo
        self.budgetRepo      = budgetRepo
        self.categoryRepo    = categoryRepo
    }

    func load(userId: String) {
        self.userId = userId
        loadSelectedMonth()
        loadMonthlyTotals()
    }

    func selectMonth(_ month: Int, year: Int) {
        selectedMonth = month
        selectedYear  = year
        loadSelectedMonth()
    }

    private func loadSelectedMonth() {
        guard !userId.isEmpty else { return }
        uiState.isLoading = true
        Task {
            let (start, end) = monthRange(month: selectedMonth, year: selectedYear)
            let dbMonth      = selectedMonth + 1   // 0-based → 1-based

            do {
                let allCats    = try categoryRepo.all()
                let catMap     = Dictionary(uniqueKeysWithValues: allCats.map { ($0.id, $0) })
                let txs        = try transactionRepo.forDateRange(userId: userId, start: start, end: end)
                let budgets    = try budgetRepo.forMonth(userId: userId, month: dbMonth, year: selectedYear)

                let expenses   = txs.filter { $0.type == .expense }
                let incomes    = txs.filter { $0.type == .income  }
                let total      = expenses.reduce(0) { $0 + $1.amount }
                let income     = incomes.reduce(0)  { $0 + $1.amount }

                let spendByCat = Dictionary(grouping: expenses, by: { $0.categoryId })
                    .mapValues { $0.reduce(0) { $0 + $1.amount } }

                let catSpends: [CategorySpend] = spendByCat
                    .compactMap { (catId, amount) in
                        guard let cid = catId, let cat = catMap[cid] else { return nil }
                        return CategorySpend(name: cat.name, icon: cat.icon, amount: amount, colorHex: cat.colorHex)
                    }
                    .sorted { $0.amount > $1.amount }

                let bars: [CategoryBudgetBar] = budgets.compactMap { budget in
                    let cat = catMap[budget.categoryId]
                    return CategoryBudgetBar(
                        categoryName: cat?.name    ?? "Other",
                        icon:         cat?.icon    ?? "📦",
                        colorHex:     cat?.colorHex ?? "#607D8B",
                        spent:        spendByCat[budget.categoryId] ?? 0,
                        minAmount:    budget.minAmount,
                        limitAmount:  budget.limitAmount
                    )
                }.sorted { $0.spent > $1.spent }

                uiState = ReportsUiState(
                    balance:             income - total,
                    totalExpense:        total,
                    totalIncome:         income,
                    transactions:        txs.sorted { $0.date > $1.date },
                    categorySpends:      catSpends,
                    categoryBudgetBars:  bars,
                    monthlyTotals:       uiState.monthlyTotals,
                    isLoading:           false
                )
            } catch {
                uiState.isLoading = false
            }
        }
    }

    private func loadMonthlyTotals() {
        guard !userId.isEmpty else { return }
        Task {
            let labels = ["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"]
            var totals: [MonthTotal] = []
            for i in stride(from: 5, through: 0, by: -1) {
                let date = Calendar.current.date(byAdding: .month, value: -i, to: .now)!
                let m    = Calendar.current.component(.month, from: date) - 1  // 0-based
                let y    = Calendar.current.component(.year,  from: date)
                let (start, end) = monthRange(month: m, year: y)
                let income  = (try? transactionRepo.totalIncome(userId: userId,  start: start, end: end)) ?? 0
                let expense = (try? transactionRepo.totalExpense(userId: userId, start: start, end: end)) ?? 0
                totals.append(MonthTotal(label: labels[m], total: income - expense))
            }
            uiState.monthlyTotals = totals
        }
    }

    private func monthRange(month: Int, year: Int) -> (Date, Date) {
        var cal = Calendar.current
        var comps = DateComponents(year: year, month: month + 1, day: 1, hour: 0, minute: 0, second: 0)
        let start = cal.date(from: comps) ?? .now
        comps.day = cal.range(of: .day, in: .month, for: start)!.count
        comps.hour = 23; comps.minute = 59; comps.second = 59
        let end = cal.date(from: comps) ?? .now
        return (start, end)
    }
}

File: android/app/src/main/java/com/budgetbuddy/ui/reports/ReportsViewModel.kt
package com.budgetbuddy.ui.reports

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.budgetbuddy.data.local.entities.TransactionEntity
import com.budgetbuddy.data.local.entities.TransactionType
import com.budgetbuddy.data.repository.BudgetRepository
import com.budgetbuddy.data.repository.CategoryRepository
import com.budgetbuddy.data.repository.TransactionRepository
import com.budgetbuddy.util.DateUtils
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import java.util.Calendar
import javax.inject.Inject

data class CategorySpend(val name: String, val icon: String, val amount: Double, val colorHex: String)

data class MonthTotal(val label: String, val total: Float)

data class CategoryBudgetBar(
    val categoryName: String,
    val icon: String,
    val colorHex: String,
    val spent: Double,
    val minAmount: Double,
    val limitAmount: Double
)

data class ReportsUiState(
    val balance: Double = 0.0,
    val totalExpense: Double = 0.0,
    val totalIncome: Double = 0.0,
    val transactions: List<TransactionEntity> = emptyList(),
    val categorySpends: List<CategorySpend> = emptyList(),
    val categoryBudgetBars: List<CategoryBudgetBar> = emptyList(),
    val monthlyTotals: List<MonthTotal> = emptyList(),
    val isLoading: Boolean = false
)

@HiltViewModel
class ReportsViewModel @Inject constructor(
    private val transactionRepository: TransactionRepository,
    private val categoryRepository: CategoryRepository,
    private val budgetRepository: BudgetRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(ReportsUiState())
    val uiState: StateFlow<ReportsUiState> = _uiState.asStateFlow()

    val categories = categoryRepository.getAllCategories()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    private var userId: String = ""
    private var selectedYear: Int = Calendar.getInstance().get(Calendar.YEAR)
    private var selectedMonth: Int = Calendar.getInstance().get(Calendar.MONTH) // 0-based
    private var monthJob: kotlinx.coroutines.Job? = null

    fun loadReports(userId: String) {
        this.userId = userId
        loadSelectedMonth()
        loadMonthlyTotals()
    }

    fun selectMonth(month: Int, year: Int) {
        selectedMonth = month
        selectedYear = year
        loadSelectedMonth()
    }

    private fun loadSelectedMonth() {
        if (userId.isEmpty()) return
        monthJob?.cancel()
        monthJob = viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            val cal = Calendar.getInstance().apply { set(selectedYear, selectedMonth, 1, 0, 0, 0); set(Calendar.MILLISECOND, 0) }
            val start = cal.timeInMillis
            cal.set(Calendar.DAY_OF_MONTH, cal.getActualMaximum(Calendar.DAY_OF_MONTH))
            cal.set(Calendar.HOUR_OF_DAY, 23); cal.set(Calendar.MINUTE, 59); cal.set(Calendar.SECOND, 59)
            val end = cal.timeInMillis

            // selectedMonth is 0-based (Calendar.MONTH); DB stores months 1-based
            val dbMonth = selectedMonth + 1

            combine(
                transactionRepository.getTransactionsByDateRange(userId, start, end),
                budgetRepository.getBudgetsForMonth(userId, dbMonth, selectedYear)
            ) { transactions, budgets ->
                val cats = categoryRepository.getAllCategories().first()
                val catMap = cats.associateBy { it.id }
                val budgetMap = budgets.associateBy { it.categoryId }

                val expenses = transactions.filter { it.type == TransactionType.EXPENSE }
                val incomes  = transactions.filter { it.type == TransactionType.INCOME }
                val totalExpense = expenses.sumOf { it.amount }
                val totalIncome  = incomes.sumOf { it.amount }

                val spendByCat = expenses.groupBy { it.categoryId }
                    .mapValues { (_, txs) -> txs.sumOf { it.amount } }

                val categorySpends = spendByCat
                    .map { (catId, amount) ->
                        val cat = catMap[catId]
                        CategorySpend(cat?.name ?: "Other", cat?.icon ?: "📦", amount, cat?.colorHex ?: "#607D8B")
                    }
                    .sortedByDescending { it.amount }

                // Only include categories that have a budget set this month
                val categoryBudgetBars = budgets.map { budget ->
                    val cat = catMap[budget.categoryId]
                    CategoryBudgetBar(
                        categoryName = cat?.name ?: "Other",
                        icon = cat?.icon ?: "📦",
                        colorHex = cat?.colorHex ?: "#607D8B",
                        spent = spendByCat[budget.categoryId] ?: 0.0,
                        minAmount = budget.minAmount,
                        limitAmount = budget.limitAmount
                    )
                }.sortedByDescending { it.spent }

                ReportsUiState(
                    balance = totalIncome - totalExpense,
                    totalExpense = totalExpense,
                    totalIncome = totalIncome,
                    transactions = transactions.sortedByDescending { it.date },
                    categorySpends = categorySpends,
                    categoryBudgetBars = categoryBudgetBars,
                    isLoading = false
                )
            }.collect { newState -> _uiState.update { newState } }
        }
    }

    private fun loadMonthlyTotals() {
        if (userId.isEmpty()) return
        viewModelScope.launch {
            val totals = mutableListOf<MonthTotal>()
            val monthLabels = listOf("Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec")
            val cal = Calendar.getInstance()
            // last 6 months
            for (i in 5 downTo 0) {
                val c = Calendar.getInstance()
                c.add(Calendar.MONTH, -i)
                val m = c.get(Calendar.MONTH)
                val y = c.get(Calendar.YEAR)
                c.set(y, m, 1, 0, 0, 0); c.set(Calendar.MILLISECOND, 0)
                val start = c.timeInMillis
                c.set(Calendar.DAY_OF_MONTH, c.getActualMaximum(Calendar.DAY_OF_MONTH))
                c.set(Calendar.HOUR_OF_DAY, 23); c.set(Calendar.MINUTE, 59); c.set(Calendar.SECOND, 59)
                val end = c.timeInMillis
                val expense = transactionRepository.getTotalExpenseForPeriod(userId, start, end)
                val income  = transactionRepository.getTotalIncomeForPeriod(userId, start, end)
                // Chart shows net balance per month (income - expense)
                totals.add(MonthTotal(monthLabels[m], (income - expense).toFloat()))
            }
            _uiState.update { it.copy(monthlyTotals = totals) }
        }
    }
}
